### PR TITLE
ARROW-6043: [Python] Array equals returns incorrectly if NaNs are in arrays

### PR DESCRIPTION
**Pull request by kszucs**
Expose `EqualOptions` and set nans_equal to True. We should probably reduce `Array::Equals` to data structure equality and move the ArrayEquals implementation to the comparison kernels.

### CODE REVIEW

File: python/pyarrow/array.pxi [hunk 1]
```diff
@@ -1009,7 +1009,8 @@ cdef class Array(_PandasConvertible):
             return NotImplemented
 
     def equals(Array self, Array other):
```
Comment (pitrou) [inline python/pyarrow/array.pxi:1011]: Why not:
```python
    def equals(Array self, Array other, *, nans_equal=True):
```

**Comment by pitrou** [comment]
Should other `equals` methods (`ChunkedArray`, `Table`, `RecordBatch`) be changed as well?

**Comment by jorisvandenbossche** [comment]
> Expose EqualOptions and set nans_equal to False.

The code actually sets it to True, not False (meaning that `equals` will return True if NaNs are at the same location)

But so that means we change the behaviour again? Because ARROW-6043 reported that it changed from True to False (which is the behaviour on master as well)

File: python/pyarrow/array.pxi (hunk 1 above)
Comment (jorisvandenbossche) [reply python/pyarrow/array.pxi:1011]: The ship might have sailed on this (I don't know how long this is already exposed in the public C++ API), but numpy has this as `equal_nan` keyword

**Comment by kszucs** [comment]
Well, the best would be to set `EqualOptions::nans_equal_` to true by default otherwise we need to wire the equal options through all calls.

**Comment by jorisvandenbossche** [comment]
@kszucs do you want to rebase/update this?

**Comment by pitrou** [comment]
This would deserve to be revived and/or restarted. @AlenkaF @raulcd Does one of you want to take it up?

**Comment by kszucs** [comment]
Due to the conflicts it would be better to restart, so I'll close this PR.

Note that wiring the default EqualOptions through the API will break the previous behaviour, so it may need a deprecation period before actually using the default settings.

**Pull request closed without merging**